Autodetect path to lxd daemon lxd socket

samuelallan72 opened the pull request:
I designed the logic by inspiration from pylxd:

https://github.com/canonical/pylxd/blob/4d4b45014d708c8097f79c2dc896116fff1fe0fa/pylxd/client.py#L393-L400

and lxd cli:

https://github.com/canonical/lxd/blob/854006f80fb9a6d5d5c42900a4352f3d0381bc13/client/connection.go#L188-L199

Fixes: #8

Commit:
Autodetect path to lxd daemon lxd socket

TheSignPainter98 commented on src/client.rs:
```diff
@@ -35,13 +36,44 @@ struct Cert {
+
+/// Return the environment variable if set and nonempty, else None.
+fn nonempty_env_var(key: &str) -> Option<String> {
+    match env::var(key) {
```
This function seems slightly problematic to have in this scope, I think it would be better to have it as a helper function inside of default until there is a clear need for it elsewhere.

Ignoring the error seems problematic as this error might not just represent an absence of the value but also that the value is present but contains invalid UTF8. This is unlikely to come up in practice, but ignoring all errors seems a little extreme. Instead, I think let's change the return type to `Result<Option<String>>` and match for `env::VarError::NotPresent` to map that to `None`, otherwise preserve the error. Perhaps in this case it should be renamed to `try_get_env_var`?

I'm also not quite sure why we need to check for empty strings, I'm not sure why the user would set `LXD_SOCKET=''` rather than just running `unset LXD_SOCKET`. I think if the user sets the empty string as their socket address, I think they'll get an error message at some point so perhaps it's okay not to check this here? It adds extra complication when describing the behaviours, so I think it'd be simpler to just drop this check

Just as an aside, I think in its current form, this function can be simplified :) Also, this makes it clearer sooner that the error is being ignored which seems nice
```rust
fn nonempty_env_var(key: &str) -> Option<String> {
    env::var(key)
        .ok()
        .and_then(|v| if v.is_empty() {
            None
        } else {
            Some(v)
        })
}

TheSignPainter98 commented on src/client.rs:
```diff
@@ -35,13 +36,44 @@ struct Cert {
-        // TODO: implement default for endpoint, autodetecting the local lxd socket
-        Endpoint::UnixSocket("/var/snap/lxd/common/lxd/unix.socket".to_owned())
+        let lxd_snap_socket_path = PathBuf::from("/var/snap/lxd/common/lxd/unix.socket");
+        let lxd_package_socket_path = PathBuf::from("/var/snap/lxd/common/lxd/unix.socket");
```
If possible, I think it would be nice to defer these allocations by instead using `Path::new` and calling `.to_owned()` if we need to convert these into a `PathBuf` more lazily.

Also, they seem to have the same string value at the moment :)

TheSignPainter98 commented on src/client.rs:
```diff
@@ -35,13 +36,44 @@ struct Cert {
+    /// 1. `LXD_SOCKET` env var if set to a nonempty string
+    /// 2. `LXD_DIR` env var if set to a nonempty string
+    /// 3. `/var/snap/lxd/common/lxd/unix.socket` (lxd snap) if path exists
+    /// 4. otherwise fall back to `/var/snap/lxd/common/lxd/unix.socket` (lxd package)
```
This is a good doc, I think it may be more visible if not attached to a trait method :)

TheSignPainter98 commented on src/client.rs:
```diff
@@ -35,13 +36,44 @@ struct Cert {
+        let lxd_snap_socket_path = PathBuf::from("/var/snap/lxd/common/lxd/unix.socket");
+        let lxd_package_socket_path = PathBuf::from("/var/snap/lxd/common/lxd/unix.socket");
+
+        let socket: PathBuf = if let Some(socket) = nonempty_env_var("LXD_SOCKET") {
```
I think this type annotation can be dropped

TheSignPainter98 commented on src/client.rs:
```diff
@@ -35,13 +36,44 @@ struct Cert {
-        Endpoint::UnixSocket("/var/snap/lxd/common/lxd/unix.socket".to_owned())
+        let lxd_snap_socket_path = PathBuf::from("/var/snap/lxd/common/lxd/unix.socket");
+        let lxd_package_socket_path = PathBuf::from("/var/snap/lxd/common/lxd/unix.socket");
+
```
The above variables seem strongly associated with the code below, so I think it would be nice to remove this newline

TheSignPainter98 reviewed (changes requested):
Looking very nice! Hopefully nothing too massive, I think the structure needs a little tweaking

Commit:
Restructure and fix logic

- fix path to lxd package unix socket
- skip checking for empty string values in env vars
- don't ignore invalid utf8 errors in env vars

TheSignPainter98 commented on src/client.rs:
```diff
@@ -35,13 +36,44 @@ struct Cert {
+    /// 2. `LXD_DIR` env var if set to a nonempty string
+    /// 3. `/var/snap/lxd/common/lxd/unix.socket` (lxd snap) if path exists
+    /// 4. otherwise fall back to `/var/snap/lxd/common/lxd/unix.socket` (lxd package)
     fn default() -> Self {
```
Perhaps `Endpoint::detect_local()`? Somehow the 'auto' part seems redundant to me. Also, I think we already have a custom `Result` type so let's make sure to use that ;)

TheSignPainter98 commented on src/client.rs:
```diff
@@ -35,13 +39,39 @@ struct Cert {
+        fn try_get_env_var(name: &str) -> Result<Option<String>> {
+            match env::var(name) {
+                Err(env::VarError::NotPresent) => Ok(None),
+                Err(x) => Err(x.into()),
+                Ok(s) => Ok(Some(s)),
```
Just some nitpicks about names:
- I'd associate `x` with a value used along the happy path, I think to highlight that this is an error, let's call the error `e` or `err`
- As we're calling `env::var`, the successful output is an environment 'var,' so I think the variable in the `Ok` case should be called `v` to match

TheSignPainter98 commented on src/client.rs:
```diff
@@ -119,15 +149,15 @@ pub struct ClientConfig {
-        Self {
-            endpoint: Endpoint::default(),
+impl ClientConfig {
+    pub fn default_autodetect_local() -> Result<Self> {
```
This name feels inconsistent with `Endpoint::autodetect_local`, I think as we're establishing a pattern it would be good to keep these the same

samuelallan72 replied:
That makes sense - I changed it to `default_with_detect_local`, to follow the convention for providing a default with a particular purpose.  What do you think?

TheSignPainter98 reviewed (changes requested):
Looking good, just a couple of nitpicks

Commit:
Update naming for clarity

samuelallan72 commented:
@TheSignPainter98 @NucciTheBoss I'll actually close my PRs here now that I've left the company and no longer have access to this repository. If I contribute in the future, I can open PRs from a personal fork. :)
